subiquity/server/server.py

```diff
@@ -294,6 +299,15 @@ def __init__(self, opts, block_log_dir):
                            self._proxy_set)
         self.geoip = GeoIP(self)
 
+        if opts.dry_run:
+            path = os.getenv("DRY_RUN_SYSTEMD_RUN_PATH")
```

Comment:
I agree with Dan here.

Reply:
As of today the only known use case is system setup on WSL, this is a really narrow and specific case.
Furthermore, the purpose of the code that uses this variable is solely for testing and very limited.

I agree that adding an option could cover other usages for the subiquity code base without systemd but AFAIK there are no such cases yet. Moreover, to implement such an option we'd have to entirely disable the code paths that use systemd, the impact much greater and out of the scope of the autoinstall mode only.
 
Furthermore, for testing, like JP is doing for the desktop installer, we'd have to restart the server in the middle of the test suite which is far less convenient than just touch'ing or rm'ing a file to simulate a system with or without a running systemd.

Pragmatically using this variable makes the change much smaller and contained than modifying the entire code base related to systemd.

Reply:
Let's take a step or two back here. I hope we can solve the issue in a way that makes all of subiquity nicer, without having to have any of this slightly convoluted logic.

I have a general question about how autoinstall for WSL operates at a user level. How is the data provided? What does the client part of system_setup even do in this situation? Does it just print a message saying, you system is ready for you, or does it just exit and it's as if it was never there?

In any case, this is about how the client knows to exit when there's an autoinstall going on. For the installer this doesn't really matter than much for "real" usage, because at the end of an autoinstall the server reboots the machine anyway so the client is going away soon whatever it does. It matters for dry-runs and so we have this odd hack where the client watches for an journald entry that indicates the server is done and exits when it sees this.

For WSL this both doesn't work (because no systemd) and actually matters (because the machine is not rebooted at the end of autoinstall) and further, because the "autoinstall" is actually very quick, it's not obvious at all on the client side whether failure to connect to the server means the server hasn't started or has finished.

So anyway, I think it makes sense to have a filesystem based backchannel for the server state always. So the server should always write the state to /run/subiquity/state or something whenever it is updated and the client can check this file in noninteractive_watch_app_state if the request fails and exit if the state is suitably terminal (we might want to add an "EXITED" state to the ApplicationState enum). Then the check in subiquity_event_noninteractive can go, and WSL should work without modification here.

Does that make sense? Have I missed any cases? I guess I'd prefer this be done in a separate branch but it might only be a small change (I won't be able to get to it until next week)

Reply:
The general operation of autoinstall for WSL is as follow:
The "preseed" file is passed as an argument to the launcher in Windows which in turn passes to it to the command line of the system setup. system setup executes as usual in autoinstall mode, there is no real difference with subiquity here apart from the required data to proceed with autoinstall. At the end of installation there is currently no message but there may be a welcome message to say that the installation is complete and ready to use. I hope this clarifies the use case.
Note that when systemd is installed but another init is running, the journal doesn't fail but is always empty and will never pass event to the client.
The logic that you describe is the one implemented in this PR which is activated when systemd is not running.

How do you want to proceed from here and move forward with this PR to enable autoinstall on WSL?

Reply:
Well I think if we merge https://github.com/canonical/subiquity/pull/1054 then we can make this branch simpler?

Reply:
See comment https://github.com/canonical/subiquity/pull/1054#pullrequestreview-758342647

Reply:
Deprecated since @mwhudson’s branch.